autoscaling/V2 for HorizontalPodAutoscaler, servicesGeneral support and serviceAccount name fix

s-silvius opened the pull request:
- Added support for autoscaling/V2 api for HPA. Change that also to default. Keep supporting autoscaling/v2beta1
- added support for serviceGeneral to able to add common labels and annotations to all services

s-silvius committed:
Add support autoscaling/v2 API to HorizontalPodAutoscaler

Change also default apiVersion to autoscaling/v2 but support still
autoscaling/v2beta1.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Add support of servicesGeneral for services.

Added
servicesGeneral.labels
servicesGeneral.annotations

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Fix serviceAccountName for the pod

serviceAccountName in pod did not matched the given name in values as ServiceAccount resource in using
helpers.app.fullname template

```
name: {{ include "helpers.app.fullname" (dict "name" $name "context" $) }}
```

So, use same template in pod to reference serviceAccount name to make them match.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Fix previous HPA change

Conditional part used template that did not existed.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Aupport for generic lifecycle, startupProbe, readinessProbe and livenessProbe

Container specific properties are merged with generic ones so that
specific ones will override generic properties.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Add support for common parameters parentChart.name and parentChart.version

When nxs-universal-chart used as library chart as dependency, parent chart name and version can be used to
override following values in labels:

app.kubernetes.io/name
helm.sh/chart
app.kubernetes.io/version

instead of nxs chart values.
`app.kubernetes.io/name` is important to be override as used as part of selectorLabels and this way parent chart's name is used in each resource in addition to release name. Without this override this label get release name as the value, i.e. value is same as for the other label `app.kubernetes.io/instance` in the selector.

If both parentChart.name and parentChart.version is given `helm.sh/chart` will be using those values instead of Chart.Name and Chart.Version whihc are values of nxs chart.

For `app.kubernetes.io/version` label the value from parentChart.version is used if given instead of Chart.Appversion or Chart.version.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

lgg42 commented:
Looking so forward to have this one merged!! :heart_eyes:

s-silvius committed:
Merge branch 'nixys:main' into main

s-silvius committed:
Added support for serviceAccount to have multiple Roles and ClusterRoles

Previously only single Role and ClusterRole was possible per serviceAccount
as role/clusterrole was a map entry.
Changed to both to list allows multiple roles/clusterroles.
Rules under each role/clusterrole are same as earlier.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Add support for RoleBinding to bind also ClusterRole

It's possible and normal valid use case to bind ClusterRole
with RoleBinding to allow clusterlevel privileges to namespace level.

Added new `clusterScope` boolean parameter for role object and if value is true
ClusterRole is used in releRef, otherwice Role is used.
Default value is false.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Merge branch 'nixys:main' into main

s-silvius committed:
Support for generic.podSecurityContext and generic.containerSecurityContext

Generic new properties if defined will add securityContext by default to all workloads
into pod level or container level respectively.

User defined securityContext on those workload resources will override these generic definitions.
But it's also possible to merge with generic definition by adding `securityContext.mergeWithGeneric`
with value `true`.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Merge remote-tracking branch 'upstream/main'

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

amontalban commented:
Thank you @s-silvius I am using this chart and found out that the HPA does not work in recent Kubernetes versions (1.32), this PR fix this.

@r-emelyanov-nixys  can we get this PR reviewed, please 🙏 ?

s-silvius committed:
Improve tpl rendering support and Openshift support for securityContext

1) Add support to omit "runAsUser" "runAsGroup" "fsGroup" from securityContext in Openshift

If
  openshift:
    securityContextConstraintsCompatibility: true
securityContext templating detects if target cluster is Openshift and omits "runAsUser" "runAsGroup" "fsGroup".
This is because Opneshift autoassigns based on SCC user/group id per namespace ranges.

kind: Namespace
apiVersion: v1
metadata:
  annotations:
    openshift.io/sa.scc.mcs: 's0:c28,c22'
    openshift.io/sa.scc.supplemental-groups: 1000800000/10000
    openshift.io/sa.scc.uid-range: 1000800000/10000

and chart defined user/group ids are not accepted.
This is inspired from Bitnami charts
https://github.com/bitnami/charts/issues/24251
https://github.com/bitnami/charts/blob/5976be1ecb807f0fd3a13b8a889da4054d7c84bf/bitnami/common/templates/_compatibility.tpl#L28

2) Improve tpl rendering to support integer and boolean values

E.g. values file you have

securityContext:
  runAsUser: 1101

deployments:
  my-app:
    securityContext:
	  runAsUser: "{{ $.Values.securityContext.runAsUser }}"

Wiht added template "helpers.tplvalues.renderAndCoerce" used nw in "helpers.securityContext" it will template into

      securityContext:
        runAsUser: 1101

instead of

      securityContext:
        runAsUser: "1101"
which would be not accepted by Kubernetes.

usage of new template "helpers.tplvalues.renderAndCoerce" added also to
_pod.tpl ("helpers.pod") -> for ports
svc.yml -> for ports

3) Improve tpl rendering support for configmaps

In _configmaps.tpl improved helper templates

"helpers.configmaps.renderConfigMap"
"helpers.configmaps.includeEnv"
"helpers.configmaps.includeEnvConfigmap"

to support better values templating.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Add automountServiceAccountToken configurability

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Add automountServiceAccountToken configurability

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Merge branch 'automountServiceAccountToken' of github.com:s-silvius/nxs-universal-chart into automountServiceAccountToken

s-silvius committed:
Add projected volume support and templating replicas

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Add projected volume support and templating replicas

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Merge branch 'projected-volume-and-replicas' of github.com:s-silvius/nxs-universal-chart into projected-volume-and-replicas

s-silvius committed:
Add ServiceAccount imagePullSecrets defaults

Add serviceAccountDefaultImagePullSecretName plus per-SA and global imagePullSecrets controls to attach registry pull secrets at the ServiceAccount level.
Update README tables/docs and harden ServiceAccount templating for serviceAccountGeneral: null and helper label/version rendering.

Signed-off-by: Samuli Silvius <samuli.silvius@nokia.com>

s-silvius committed:
Merge pull request #1 from s-silvius/image-pull-secret

Add ServiceAccount imagePullSecrets defaults

RukInDaHouse commented:
Реализовано в PR #106
